src/NuGet.Core/NuGet.Commands/SignCommand/CertificateProvider.cs

```diff
@@ -132,23 +134,33 @@ internal static class CertificateProvider
 
         private static X509Certificate2Collection LoadCertificateFromStore(CertificateSourceOptions options)
         {
-            var resultCollection = new X509Certificate2Collection();
+            X509Certificate2Collection resultCollection = null;
             var store = new X509Store(options.StoreName, options.StoreLocation);
 
             OpenStore(store);
 
+            // Passing true for validOnly seems like a good idea; it would filter out invalid certificates.
+            // However, "invalid certificates" is a broad category that includes untrusted self-issued certificates.
+            // Untrusted self-issued certificates are permitted at signing time, so we must perform certificate
+            // validity checks ourselves.
+            const bool validOnly = false;
+
             if (!string.IsNullOrEmpty(options.Fingerprint))
             {
-                resultCollection = store.Certificates.Find(X509FindType.FindByThumbprint, options.Fingerprint, validOnly: true);
+                resultCollection = store.Certificates.Find(X509FindType.FindByThumbprint, options.Fingerprint, validOnly);
             }
             else if (!string.IsNullOrEmpty(options.SubjectName))
             {
-                resultCollection = store.Certificates.Find(X509FindType.FindBySubjectName, options.SubjectName, validOnly: true);
+                resultCollection = store.Certificates.Find(X509FindType.FindBySubjectName, options.SubjectName, validOnly);
             }
 
 #if IS_DESKTOP
             store.Close();
 #endif
+
+            resultCollection = resultCollection ?? new X509Certificate2Collection();
+            resultCollection = GetValidCertificates(resultCollection);
```

Comment:
Will this filter out self-issued certificates?

Reply:
Not purely on the basis of them being self-issued.  See [here](https://github.com/NuGet/NuGet.Client/blob/21409dd7f526c87d4d6a2a91657b3dbc3fb53353/src/NuGet.Core/NuGet.Packaging/Signing/Utility/CertificateChainUtility.cs#L143).  If a self-issued certificate is invalid for another reason it will be filtered out.